### PR TITLE
[feature][broker] PIP 37: Support chunking with Shared subscription

### PR DESCRIPTION
**Pull request by BewareMyPower**
### Motivation

https://github.com/apache/pulsar/wiki/PIP-37%3A-Large-message-size-handling-in-Pulsar#option-1-broker-caches-mapping-of-message-uuid-and-consumerid

The cons of option 1 described in the original proposal don't exist
for current code because broker keeps redelivered messages into
**sorted** map now.

### Modifications

First of all, to avoid too many code changes, an `EntryAndMetadata`
class is introduced to bind the `Entry` with the associated
`MessageMetadata` to avoid parsing the metadata repeatedly. It also
implements the `Entry` interface, so this PR changes some
`List<Entry>` parameters to `List<? extends Entry>` so that a
`List<EntryAndMetadata>` argument can be accepted.

Then, a `SharedConsumerAssignor` is introduced to assign a list of
entries to all shared consumers.
1. Use a default selector to select the next consumer, like
   `PersistentDispatcherMultipleConsumers#getNextConsumer`,
2. Each time a consumer is chosen, assign the entries in range
   [i, i+permits) to the consumer except entries that have uuid:
  - If uuid is not cached, cache `uuid -> consumer` to indicate the
    chunked message of this uuid must be dispatched to this consumer.
  - Otherwise, assign this entry to the owner consumer of the uuid.

The `assign` method returns a map that maps `Consumer` to
`List<EntryAndMetadata>`. The following logic is similar to the
Key_Shared dispatcher.

Finally, cancel the limit in `ConsumerImpl`.

### Verifying this change

`SharedConsumerAssignorTest` is added to show how the assignor works
in detail.

`MessageChunkingSharedTest` is added to verify the Shared dispatcher
works on chunked messages, including:
- Single producer sends chunked messages with various chunk count to a
  consumer has a limited permits.
- Single producer sends chunked messages to two consumers to verify
  both they can receive chunked messages.
- Produce interleaved chunks via `PersistentTopic` directly to simulate
  multiple producers, and verify the new consumer can receive all
  unacknowledged messages received by the old consumer.

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java [hunk 1]
```diff
@@ -129,6 +133,8 @@ public PersistentDispatcherMultipleConsumers(PersistentTopic topic, ManagedCurso
                 : RedeliveryTrackerDisabled.REDELIVERY_TRACKER_DISABLED;
         this.readBatchSize = serviceConfig.getDispatcherMaxReadBatchSize();
         this.initializeDispatchRateLimiterIfNeeded();
+        this.assignor = new SharedConsumerAssignor(this::getNextConsumer,
+                entry -> addMessageToReplay(entry.getLedgerId(), entry.getEntryId()));
```
Comment (RobertIndie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:137]: Should we release this entry after addMessageToReplay?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java [hunk 2]
```diff
@@ -607,19 +645,72 @@ protected void sendMessagesToConsumers(ReadType readType, List<Entry> entries) {
...
-                log.debug("[{}] No consumers found with available permits, storing {} positions for later replay", name,
-                        entries.size() - start);
+                log.debug("[{}] select consumer {} with messages num {}, read type is {}",
+                        name, consumer.consumerName(), messagesForC, readType);
             }
-            entries.subList(start, entries.size()).forEach(entry -> {
-                long stickyKeyHash = getStickyKeyHash(entry);
-                addMessageToReplay(entry.getLedgerId(), entry.getEntryId(), stickyKeyHash);
-                entry.release();
+            if (messagesForC < entryAndMetadataList.size()) {
```
Comment (RobertIndie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:672]: When will messagesForC or `consumer.getAvailablePermits()` be less than `entryAndMetadataList.size()`?

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingSharedTest.java [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        final Set<String> values = new HashSet<>();
+        for (int i = 0; i < 10; i++) {
+            values.add(createChunkedMessage(4));
+        }
+        for (String value : values) {
+            producer.send(value);
+        }
+
+        Awaitility.await().atMost(Duration.ofSeconds(3))
+                .until(() -> receivedValues1.size() + receivedValues2.size() >= values.size());
```
Comment (RobertIndie) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingSharedTest.java:136]: Maybe check if they are equal is enough here?

**Review by RobertIndie — COMMENTED** [review]
Overall looks good to me. Left some comments.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java (hunk 2 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:672]: I'm not sure, maybe the consumer's permits might change in the loop. It's added here for safety programming.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingSharedTest.java (hunk 3 above)
Comment (BewareMyPower) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChunkingSharedTest.java:136]: No. If the consumer side doesn't merge the chunks (the wrong behavior), the total size could be greater. And yes, I will add an assertion to verify they are equal, which should be the correct behavior.

**Comment by RobertIndie** [comment]
This PR will fixes https://github.com/apache/pulsar/issues/7645

**Review by RobertIndie — APPROVED** [review]

**Comment by BewareMyPower** [comment]
@codelipenghui @merlimat @rdhabalia @Jason918 @eolivelli @315157973 Could you take a look at this PR?

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/SharedConsumerAssignorTest.java [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import org.apache.bookkeeper.mledger.impl.EntryImpl;
+import org.apache.pulsar.common.allocator.PulsarByteBufAllocator;
+import org.apache.pulsar.common.api.proto.MessageMetadata;
+import org.apache.pulsar.common.protocol.Commands;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+public class SharedConsumerAssignorTest {
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/SharedConsumerAssignorTest.java:43]: Need to add the test group.

**Comment by codelipenghui** [comment]
@BewareMyPower Please help resolve the conflicts.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/SharedConsumerAssignorTest.java (hunk 4 above)
Comment (BewareMyPower) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/SharedConsumerAssignorTest.java:44]: Done.

**Pull request merged**